cypher-shell/src/main/java/org/neo4j/shell/parser/ShellStatementParser.java

```diff
@@ -0,0 +1,188 @@
+package org.neo4j.shell.parser;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+/**
+ * A cypher aware parser which can detect shell commands (:prefixed) or cypher.
+ */
+public class ShellStatementParser implements StatementParser {
+
+    private static final Pattern shellCmdPattern = Pattern.compile("^\\s*:.+\\s*$");
+    private static final char SEMICOLON = ';';
+    private static final char BACKSLASH = '\\';
+    private static final String LINE_COMMENT_START = "//";
+    private static final String LINE_COMMENT_END = "\n";
+    private static final String BLOCK_COMMENT_START = "/*";
+    private static final String BLOCK_COMMENT_END = "*/";
+    private static final char BACKTICK = '`';
+    private static final char DOUBLE_QUOTE = '"';
+    private static final char SINGLE_QUOTE = '\'';
+    private Optional<String> awaitedRightDelimiter;
+    private StringBuilder statement;
+    private ArrayList<String> parsedStatements;
+
+
+    public ShellStatementParser() {
+        parsedStatements = new ArrayList<>();
+        statement = new StringBuilder();
+        awaitedRightDelimiter = Optional.empty();
+    }
+
+    @Override
+    public void parseMoreText(@Nonnull String line) {
+        // See if it could possibly be a shell command, only valid if not in a current statement
+        if (!statementStarted() && shellCmdPattern.matcher(line).find()) {
```

Comment:
This method is used in negation it can instead read as statementNotStarted and the method implementation could be changed.
```
private boolean statementStarted() {
        return !statement.toString().trim().isEmpty();
    }

```